src/ucp/proto/proto_multi.c

```diff
@@ -33,103 +33,48 @@ ucp_proto_multi_compare_tl_perf(const void *elem1, const void *elem2, void *arg)
-        (params->max_lanes == 0)) {
-        return UCS_ERR_UNSUPPORTED;
-    }
-
-    /* Find first lane */
-    num_lanes = ucp_proto_common_find_lanes_with_min_frag(
-            &params->super, params->first.lane_type, params->first.tl_cap_flags,
-            1, 0, lanes);
-    if (num_lanes == 0) {
-        ucs_trace("no lanes for %s",
-                  ucp_proto_id_field(params->super.super.proto_id, name));
-        return UCS_ERR_NO_ELEM;
-    }
-
-    /* Find rest of the lanes */
-    num_lanes += ucp_proto_common_find_lanes_with_min_frag(
-            &params->super, params->middle.lane_type,
-            params->middle.tl_cap_flags, UCP_PROTO_MAX_LANES - 1,
-            UCS_BIT(lanes[0]), lanes + 1);
+    ucp_lane_map_t lane_map = 0;
+    ucp_proto_common_tl_perf_t *lane_perf;
+    ucp_lane_index_t i, lane;
 
-    /* Get bandwidth of all lanes and max_bandwidth */
-    max_bandwidth = 0;
     for (i = 0; i < num_lanes; ++i) {
         lane      = lanes[i];
         lane_perf = &lanes_perf[lane];
 
-        status = ucp_proto_common_get_lane_perf(&params->super, lane, lane_perf,
-                                                &lanes_perf_nodes[lane]);
-        if (status != UCS_OK) {
-            return status;
+        if (ucs_popcount(lane_map) == max_lanes) {
+            break;
         }
-
-        /* Calculate maximal bandwidth of all lanes, to skip slow lanes */
-        max_bandwidth = ucs_max(max_bandwidth, lane_perf->bandwidth);
-    }
-
-    /* Sort lanes by bandwidth */
-    if (num_lanes > 1) {
-        /* When the first lane is different from the middle one, then we should
-         * preserve the ordering with eager protocols for TAG and AM APIs */
-        first_lane = (params->first.lane_type == params->middle.lane_type) ? 0:
-                                                                             1;
-        ucs_qsort_r(lanes + first_lane, num_lanes - first_lane,
-                    sizeof(ucp_lane_index_t), ucp_proto_multi_compare_tl_perf,
-                    lanes_perf);
-    }
-
-    /* Select the lanes to use, and calculate their aggregate performance */
-    perf.bandwidth          = 0;
-    perf.send_pre_overhead  = 0;
-    perf.send_post_overhead = 0;
-    perf.recv_overhead      = 0;
-    perf.latency            = 0;
-    perf.sys_latency        = 0;
-    lane_map                = 0;
-    max_frag_ratio          = 0;
-    min_bandwidth           = DBL_MAX;
-
-    for (i = 0; (i < num_lanes) && (ucs_popcount(lane_map) < params->max_lanes);
-         ++i) {
-        lane      = lanes[i];
-        lane_perf = &lanes_perf[lane];
         if ((lane_perf->bandwidth * max_bw_ratio) < max_bandwidth) {
             /* Bandwidth on this lane is too low compared to the fastest
                available lane, so it's not worth using it */
             continue;
         }
+        lane_map |= UCS_BIT(lane);
+    }
+
+    return lane_map;
+}
+
+static ucp_proto_common_tl_perf_t
```

Comment:
pass some params by const
Returning by value is intentional?

Reply:
Done, thanks

Yes, it is intentional